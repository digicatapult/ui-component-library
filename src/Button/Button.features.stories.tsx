import {EyeIcon, HeartIcon} from '@primer/octicons-react'
import React, {useState} from 'react'
import {Button} from '.'

export default {
  title: 'Components/Button/Features',
}

export const Default = () => <Button>Default</Button>

export const Primary = () => <Button variant="primary">Primary</Button>

export const Danger = () => <Button variant="danger">Danger</Button>

export const Invisible = () => <Button variant="invisible">Invisible</Button>

export const Outline = () => <Button variant="outline">Outline</Button>

export const LeadingVisual = () => <Button leadingIcon={HeartIcon}>Leading visual</Button>

export const TrailingVisual = () => <Button trailingIcon={EyeIcon}>Trailing visual</Button>

export const TrailingCounter = () => {
  const [count, setCount] = useState(0)
  return (
    <Button onClick={() => setCount(count + 1)}>
      Watch
      <Button.Counter>{count}</Button.Counter>
    </Button>
  )
}

export const Disabled = () => <Button disabled>Default</Button>

export const Small = () => <Button size="small">Default</Button>

export const Medium = () => <Button size="medium">Default</Button>

export const Large = () => <Button size="large">Default</Button>
