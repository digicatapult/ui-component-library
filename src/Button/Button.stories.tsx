import React from 'react'
import {EyeClosedIcon, EyeIcon, SearchIcon, XIcon, HeartIcon} from '@primer/octicons-react'
import {Story, Meta} from '@storybook/react'
import {Button} from '.'
import {OcticonArgType} from '../utils/story-helpers'

export default {
  title: 'Components/Button',
  argTypes: {
    size: {
      control: {
        type: 'radio',
      },
      options: ['small', 'medium', 'large'],
    },
    disabled: {
      control: {
        type: 'boolean',
      },
    },
    variant: {
      control: {
        type: 'radio',
        options: ['default', 'primary', 'danger', 'invisible', 'outline'],
      },
    },
    leadingIcon: OcticonArgType([EyeClosedIcon, EyeIcon, SearchIcon, XIcon, HeartIcon]),
    trailingIcon: OcticonArgType([EyeClosedIcon, EyeIcon, SearchIcon, XIcon, HeartIcon]),
  },
  args: {
    size: 'medium',
    disabled: false,
    variant: 'default',
    trailingIcon: null,
    leadingIcon: null,
  },
} as Meta<typeof Button>

export const Playground: Story<typeof Button> = args => <Button {...args}>Default</Button>
