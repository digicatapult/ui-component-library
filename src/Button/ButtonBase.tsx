import React, {ComponentPropsWithRef, forwardRef, useMemo} from 'react'
import {ForwardRefComponent as PolymorphicForwardRefComponent} from '../utils/polymorphic'
import Box from '../Box'
import {merge, SxProp} from '../sx'
import {useTheme} from '../ThemeProvider'
import {ButtonProps, StyledButton} from './types'
import {getVariantStyles, getSizeStyles, getButtonStyles} from './styles'
import {useRefObjectAsForwardedRef} from '../hooks/useRefObjectAsForwardedRef'
declare let __DEV__: boolean

const defaultSxProp = {}
const iconWrapStyles = {
  display: 'inline-block',
}
const trailingIconStyles = {
  ...iconWrapStyles,
  ml: 2,
}

const ButtonBase = forwardRef<HTMLElement, ButtonProps>(
  ({children, as: Component = 'button', sx: sxProp = defaultSxProp, ...props}, forwardedRef): JSX.Element => {
    const {leadingIcon: LeadingIcon, trailingIcon: TrailingIcon, variant = 'default', size = 'medium', ...rest} = props
    const innerRef = React.useRef<HTMLElement>(null)
    useRefObjectAsForwardedRef(forwardedRef, innerRef)

    const {theme} = useTheme()
    const baseStyles = useMemo(() => {
      return merge.all([getButtonStyles(theme), getSizeStyles(size, variant, false), getVariantStyles(variant, theme)])
    }, [theme, size, variant])
    const sxStyles = useMemo(() => {
      return merge(baseStyles, sxProp as SxProp)
    }, [baseStyles, sxProp])

    if (__DEV__) {
      /**
       * The Linter yells because it thinks this conditionally calls an effect,
       * but since this is a compile-time flag and not a runtime conditional
       * this is safe, and ensures the entire effect is kept out of prod builds
       * shaving precious bytes from the output, and avoiding mounting a noop effect
       */
      // eslint-disable-next-line react-hooks/rules-of-hooks
      React.useEffect(() => {
        if (!(innerRef.current instanceof HTMLButtonElement) && !(innerRef.current instanceof HTMLAnchorElement)) {
          // eslint-disable-next-line no-console
          console.warn('This component should be an instanceof a semantic button or anchor')
        }
      }, [innerRef])
    }

    return (
      <StyledButton as={Component} sx={sxStyles} {...rest} ref={innerRef}>
        {LeadingIcon && (
          <Box as="span" data-component="leadingIcon" sx={iconWrapStyles}>
            <LeadingIcon />
          </Box>
        )}
        {children && <span data-component="text">{children}</span>}
        {TrailingIcon && (
          <Box as="span" data-component="trailingIcon" sx={trailingIconStyles}>
            <TrailingIcon />
          </Box>
        )}
      </StyledButton>
    )
  },
) as PolymorphicForwardRefComponent<'button' | 'a', ButtonProps>

export type ButtonBaseProps = ComponentPropsWithRef<typeof ButtonBase>

export {ButtonBase}
