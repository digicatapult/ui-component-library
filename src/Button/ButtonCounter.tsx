import React from 'react'
import {SxProp} from '../sx'
import CounterLabel from '../CounterLabel'

export type CounterProps = {
  children: number
} & SxProp

const Counter = ({children, sx: sxProp = {}, ...props}: CounterProps) => {
  return (
    <CounterLabel data-component="ButtonCounter" sx={{ml: 2, ...sxProp}} {...props}>
      {children}
    </CounterLabel>
  )
}

export {Counter}
