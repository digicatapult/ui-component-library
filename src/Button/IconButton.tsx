import React, {forwardRef} from 'react'
import {merge, SxProp} from '../sx'
import {useTheme} from '../ThemeProvider'
import Box from '../Box'
import {IconButtonProps, StyledButton} from './types'
import {getBaseStyles, getSizeStyles, getVariantStyles} from './styles'

const IconButton = forwardRef<HTMLButtonElement, IconButtonProps>((props, forwardedRef): JSX.Element => {
  const {variant = 'default', size = 'medium', sx: sxProp = {}, icon: Icon, ...rest} = props
  const {theme} = useTheme()
  const sxStyles = merge.all([
    getBaseStyles(theme),
    getSizeStyles(size, variant, true),
    getVariantStyles(variant, theme),
    sxProp as SxProp,
  ])
  return (
    <StyledButton sx={sxStyles} {...rest} ref={forwardedRef}>
      <Box as="span" sx={{display: 'inline-block'}}>
        <Icon />
      </Box>
    </StyledButton>
  )
})

export {IconButton}
