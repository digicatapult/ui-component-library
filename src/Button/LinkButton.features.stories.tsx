import {EyeIcon, HeartIcon} from '@primer/octicons-react'
import React, {forwardRef} from 'react'
import {Button} from '.'

export default {
  title: 'Components/LinkButton/Features',
}

export const Default = () => (
  <Button as="a" href="/">
    Default
  </Button>
)

export const Primary = () => (
  <Button as="a" href="/" variant="primary">
    Primary
  </Button>
)

export const Danger = () => (
  <Button as="a" href="/" variant="danger">
    Danger
  </Button>
)

export const Invisible = () => (
  <Button as="a" href="/" variant="invisible">
    Invisible
  </Button>
)

export const Outline = () => (
  <Button as="a" href="/" variant="outline">
    Invisible
  </Button>
)

export const LeadingVisual = () => (
  <Button as="a" href="/" leadingIcon={HeartIcon}>
    Leading visual
  </Button>
)

export const TrailingVisual = () => (
  <Button as="a" href="/" trailingIcon={EyeIcon}>
    Trailing visual
  </Button>
)

export const Small = () => (
  <Button as="a" href="/" size="small">
    Default
  </Button>
)

export const Medium = () => (
  <Button as="a" href="/" size="medium">
    Default
  </Button>
)

export const Large = () => (
  <Button as="a" href="/" size="large">
    Default
  </Button>
)

type ReactRouterLikeLinkProps = {to: string; children: React.ReactNode}
const ReactRouterLikeLink = forwardRef<HTMLAnchorElement, ReactRouterLikeLinkProps>(
  ({to, ...props}: {to: string; children: React.ReactNode}, ref) => {
    // eslint-disable-next-line jsx-a11y/anchor-has-content
    return <a ref={ref} href={to} {...props} />
  },
)

export const WithReactRouter = () => (
  <Button to="/dummy" as={ReactRouterLikeLink}>
    Default
  </Button>
)
