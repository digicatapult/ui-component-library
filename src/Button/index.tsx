import React from 'react'
import styled from 'styled-components'

type buttonVariant =
  | 'rounded'
  | 'square'
  | 'roundedShadow'
  | 'squareShadow'
  // NEW
  | 'roundedPronounced'
  | 'roundedPronouncedShadow'

export type ButtonTextProps = {
  styles?: React.CSSProperties
  type?: 'submit'
  name?: string
  variant?: buttonVariant
  onClick?: (e: any) => void
}

const ButtonBasic: React.FC<React.PropsWithChildren<ButtonTextProps>> = ({
  children,
  variant,
  ...props
}) => {
  let Button: React.FunctionComponent<
    React.ButtonHTMLAttributes<HTMLButtonElement>
  > | null = null

  switch (variant) {
    case 'rounded':
      Button = RoundedButton
      break
    case 'square':
      Button = SquareButton
      break
    case 'roundedShadow':
      Button = RoundedShadowButton
      break
    case 'squareShadow':
      Button = SquareShadowButton
      break
    // NEW
    case 'roundedPronounced':
      Button = RoundedPronouncedButton
      break
    case 'roundedPronouncedShadow':
      Button = RoundedPronouncedShadowButton
      break

    default:
      Button = RoundedShadowButton
      break
  }

  return <Button {...props}>{children}</Button>
}

const SquareButton = styled.button`
  height: 30px;
  cursor: pointer;
  border: 1px solid #000;
  padding: 0px 15px;
  background: #fff;
`

const RoundedButton = styled(SquareButton)`
  border-radius: 8px;
`

const SquareShadowButton = styled(SquareButton)`
  box-shadow: 0px 2px 0px 0px #000;
`

const RoundedShadowButton = styled(RoundedButton)`
  box-shadow: 0px 2px 0px 0px #000;
`

// NEW

const RoundedPronouncedButton = styled(SquareButton)`
  border-radius: 4em;
`

const RoundedPronouncedShadowButton = styled(RoundedPronouncedButton)`
  box-shadow: 0px 2px 0px 0px #000;
`

export default ButtonBasic
