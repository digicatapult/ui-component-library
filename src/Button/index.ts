import {ButtonComponent} from './Button'
import {Counter} from './ButtonCounter'
import {IconButton} from './IconButton'
import {LinkButton} from './LinkButton'
export type {ButtonProps, IconButtonProps} from './types'
export {IconButton, LinkButton}

export const Button = Object.assign(ButtonComponent, {
  Counter,
})
