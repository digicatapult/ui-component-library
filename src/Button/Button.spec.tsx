import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Button from './index.js'

describe('Input', () => {
  test('submit', () => {
    const instance = renderer.create(<Button type="submit">press me</Button>)
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('default', () => {
    const instance = renderer.create(<Button name="default">default</Button>)
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('rounded', () => {
    const instance = renderer.create(
      <Button name="default" variant="rounded">
        default
      </Button>,
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('square', () => {
    const instance = renderer.create(
      <Button name="default" variant="square">
        default
      </Button>,
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('roundedShadow', () => {
    const instance = renderer.create(
      <Button name="default" variant="roundedShadow">
        default
      </Button>,
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('squareShadow', () => {
    const instance = renderer.create(
      <Button name="default" variant="squareShadow">
        default
      </Button>,
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('roundedPronounced', () => {
    const instance = renderer.create(
      <Button name="default" variant="roundedPronounced">
        default
      </Button>,
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('roundedPronouncedShadow', () => {
    const instance = renderer.create(
      <Button name="default" variant="roundedPronounced">
        default
      </Button>,
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('style overrides', () => {
    const instance = renderer.create(
      <Button name="override-a" styles={{ height: '100px', background: 'red' }}>
        press me
      </Button>,
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
