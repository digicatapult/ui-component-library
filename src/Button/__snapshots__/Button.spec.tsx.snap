// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input default 1`] = `
.c0 {
  height: 30px;
  border-radius: 8px;
  border: 1px solid #000;
  padding: 0px 15px;
  background: #fff;
  margin-right: 3px;
  box-shadow: 0px 2px 0px 0px #000;
}

<button
  className="c0"
  name="default"
>
  default
</button>
`;

exports[`Input style overrides 1`] = `
.c0 {
  height: 30px;
  border-radius: 8px;
  border: 1px solid #000;
  padding: 0px 15px;
  background: #fff;
  margin-right: 3px;
  box-shadow: 0px 2px 0px 0px #000;
}

<button
  className="c0"
  name="override-a"
  styles={
    {
      "background": "red",
      "height": "100px",
    }
  }
>
  press me
</button>
`;

exports[`Input submit 1`] = `
.c0 {
  height: 30px;
  border-radius: 8px;
  border: 1px solid #000;
  padding: 0px 15px;
  background: #fff;
  margin-right: 3px;
  box-shadow: 0px 2px 0px 0px #000;
}

<button
  className="c0"
  type="submit"
>
  press me
</button>
`;
